tools/package-source.js

```diff
@@ -1223,20 +1227,93 @@ _.extend(PackageSource.prototype, {
 
         // Read top-level subdirectories. Ignore subdirectories that have
         // special handling.
+        var excludedTopLevelSubdirectories = [/^packages\/$/,
+          // XXX We no longer actually have special handling
+          //     for the programs subdirectory, but let's not
+          //     suddenly start treating it as part of the main
+          //     app program.
+          /^programs\/$/,
+          /^public\/$/, /^private\/$/,
+          /^cordova-build-override\/$/,
+          otherUnibuildRegExp].concat(sourceExclude)
+        if (!options.includeTests) {
+          excludedTopLevelSubdirectories.push(/^tests\/$/);
+        }
         var sourceDirectories = readAndWatchDirectory('', {
           include: [/\/$/],
-          exclude: [/^packages\/$/, /^tests\/$/,
-                    // XXX We no longer actually have special handling
-                    //     for the programs subdirectory, but let's not
-                    //     suddenly start treating it as part of the main
-                    //     app program.
-                    /^programs\/$/,
-                    /^public\/$/, /^private\/$/,
-                    /^cordova-build-override\/$/,
-                    otherUnibuildRegExp].concat(sourceExclude)
+          exclude: excludedTopLevelSubdirectories
         });
         checkForInfiniteRecursion('');
 
+        // Inputs that should return true:
+        // tests/...
+        // .../tests/...
+        var isTestsSubdirectory = function (dir) {
```

Comment:
Can we factor these out into a file that has to do with testing? I'd rather not make this function even longer than it is now - 500 lines is really pushing it.